app/helpers/pagelets_helper.rb

```diff
@@ -67,8 +67,70 @@ def th_content(col)
     sort col[:key], as: col[:label], default: col[:default_sort] || 'ASC'
   end
 
+  def mount_column_selector
+    url = "#{api_users_path}/#{User.current[:id]}/table_preferences"
+    has_preference = !@selected_columns.nil?
+    columns = columns_view
+    react_component("ColumnSelector", data:
+      {
+        url: url,
+        controller: controller_name,
+        columns: columns,
+        initialColumns: columns,
+        hasPreference: has_preference,
+      }
+    )
+  end
+
   private
 
+  def defined_columns
+    Pagelets::Manager.pagelets_at('hosts/_list', :hosts_table_column_header).map do |pt|
+      next unless pt.opts[:key]
+      {
+        key: pt.opts[:key],
+        label: pt.opts[:label],
+        profiles: pt.profiles.map { |pr| pr.id },
+        checked: @selected_columns ? @selected_columns.include?(pt.opts[:key].to_s) : pt.profiles.any? { |profile| profile.default? },
+      }
+    end.compact
+  end
+
+  def defined_categories(columns)
+    columns.map { |col| col[:profiles] }.flatten.uniq
```

Comment:
In case if `col[:profiles]` is an array of profiles instead of an array of strings, then it should be: 
```suggestion
    columns.map { |col| col[:profiles] }.flatten.uniq { |pl| pl.id }
```